// easy-currencies/dist/parts/config.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.resolveProviders = exports.Config = void 0;
const axios_1 = __importDefault(require("axios"));
const providers_1 = require("./providers");
const utils_1 = require("./utils");
/**
 * Config object that initializes with configuration data
 * passed in by the user.
 *
 * @export
 * @class Config
 */
class Config {
    /**
     * Creates an instance of Config.
     * @param {(...ProviderReference[] | undefined[] | string[])} config
     * @memberof Config
     */
    constructor(...config) {
        /**
         * Active client.
         */
        this._client = axios_1.default.create();
        /**
         * Client setter.
         * @param client  The client.
         */
        this.setClient = (client) => {
            this._client = client;
        };
        /**
         * Client getter.
         */
        this.getClient = () => {
            return this._client;
        };
        /**
         * Adds a single new, user-defined provider to the list of providers.
         *
         * @param {string} name - the new provider name
         * @param {Provider} provider - the new provider object
         * @param {boolean} [setActive=false] - should the new provider(s) be prioritized
         * @memberof Config
         */
        this.add = (name, provider, setActive = false) => {
            this.addMultiple([{ name, provider }], setActive);
        };
        /**
         * Adds multiple new, user-defined provider to the list of providers.
         *
         * @param {UserDefinedProvider[]} providers - providers to be added
         * @param {boolean} [setActive=false] - should the new provider(s) be prioritized
         * @memberof Config
         */
        this.addMultiple = (newProviders, setActive = false) => {
            // Duplicate check
            newProviders.forEach((p) => {
                if (!(0, utils_1.checkIfUserDefinedProvider)(p)) {
                    throw "Invalid provider format!";
                }
                if (providers_1.providers[p.name]) {
                    throw "A provider by this name is already registered!";
                }
                providers_1.providers[p.name] = p.provider;
            });
            // Adding provider to active providers
            this.addProviders(newProviders.map((p) => p.provider), setActive);
        };
        /**
         * Removes a specific provider
         * @param {Provider} provider - provider to be removed
         * @memberof Config
         */
        this.remove = (provider) => {
            this._active = this._active.filter((p) => p != provider);
        };
        this._active = resolveProviders(...config);
        // adding default fallback
        this.addProviders([providers_1.providers.ExchangeRateAPI], false);
    }
    /**
     * Provider getter.
     *
     * @returns {Provider[]}
     * @memberof Config
     */
    get providers() {
        return this._active;
    }
    /**
     * Provider setter (adder).
     *
     * @param {Provider[]} providers - providers to be added
     * @param {boolean} [setActive=false] - should the new provider(s) be prioritized
     * @returns {void}
     * @memberof Config
     */
    addProviders(providers, setActive) {
        providers = providers.filter((p) => {
            return !this._active.find((a) => a == p);
        });
        if (setActive) {
            this._active.unshift(...providers);
            return;
        }
        this._active.push(...providers);
        return;
    }
    /**
     * Returns the current provider
     *
     * @returns {Provider} - current provider
     * @memberof Config
     */
    activeProvider() {
        return this._active[0];
    }
}
exports.Config = Config;
/**
 * Config resolver that normalizes configuration input into the config interface
 *
 * @export
 * @param {(...ProviderReference[] | undefined[] | string[])} configuration
 * @returns {Provider[]}
 */
function resolveProviders(...configuration) {
    // resolve default if none provided.
    if (typeof configuration === "undefined" || !configuration.length) {
        return [providers_1.providers.ExchangeRateAPI];
    }
    // checking for incorrect config types
    if (typeof configuration[0] !== "object" &&
        typeof configuration[0] !== "undefined" &&
        typeof configuration[0] !== "string") {
        throw "You must either supply nothing or a config object (see the 'config' section to see the different APIs that can be used)";
    }
    // returning single provider
    if (typeof configuration[0] === "string") {
        // constructing in initializationConfig object from string values
        return [(0, providers_1.resolveProvider)({ name: configuration[0], key: configuration[1] })];
    }
    // configuration is an array of providers
    // casting
    const initializationConfig = configuration;
    // resolving all providers
    return initializationConfig.map((provider) => (0, providers_1.resolveProvider)(provider));
}
exports.resolveProviders = resolveProviders;
